File: Algorithms/0304_Range_Sum_Query_2D_-_Immutable/Python/Range_Sum_Query_2D_-_Immutable_Solution_1.py

# Space: O(1)
# Time: O(n)

class NumMatrix:

    def __init__(self, matrix):
        self.board = matrix

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        res = 0
        for y in range(row1, row2 + 1):
            for x in range(col1, col2 + 1):
                res += self.board[y][x]
        return res

# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: Algorithms/0304_Range_Sum_Query_2D_-_Immutable/Java/RangeSumQuery2D-ImmutableSolution1.java

// Space: O(1)
// Time: O(n)

public class NumMatrix {

    int[][] board;

    public NumMatrix(int[][] matrix) {
        board = matrix;
    }

    public int sumRegion(int row1, int col1, int row2, int col2) {
        int res = 0;
        for (int y = row1; y <= row2; y++) {
            for (int x = col1; x <= col2; x++) {
                res += this.board[y][x];
            }
        }
        return res;
    }
}




